solidity/contracts/liquidity-protection/TokenTimeWeightedAverage.sol

```diff
@@ -0,0 +1,375 @@
+     *
+     * @param _token the token the data is accumulated for
+     * @param _sample the sample to add
+     * @param _time the sampling timestamp
+     */
+    function addSample(
+        IERC20Token _token,
+        Fraction memory _sample,
+        uint256 _time
+    ) private {
+        TokenData storage tokenData = data[_token];
+        uint256 lastSampleTime = tokenData.lastSampleTime;
+
+        // make sure that the samples are added in the correct order
+        require(_time >= lastSampleTime, "ERR_WRONG_ORDER");
+
+        // if this is the initialization sample for this contract, set it as the first sample and return
+        if (tokenData.firstSampleTime == 0) {
+            tokenData.firstSampleTime = _time;
+            tokenData.lastSampleTime = _time;
+
+            tokenData.accumulators[_time] = _sample;
+            tokenData.timestamps[_time] = true;
+
+            emit SampleAdded(_token, _sample.n, _sample.d, _time);
+
+            return;
+        }
+
+        Fraction memory lastAccumulator = tokenData.accumulators[lastSampleTime];
+
+        // update the previous accumulator value once per-block. we would need it in order to accumulate
+        // same-block changes
+        if (_time > tokenData.prevAccumulatorUpdateTime) {
+            tokenData.prevAccumulatorUpdateTime = _time;
+            tokenData.prevAccumulator = lastAccumulator;
+        }
+
+        // if we already have a sample for this timestamp - use the backup of the previous accumulator and use it in
+        // combination with the new sample
+        if (_time == lastSampleTime) {
+            lastAccumulator = tokenData.prevAccumulator;
+        }
+
+        // accumulate the current value in combination with the last accumulator using the TWA formuls:
+        //  ACC[time] = ACC[time - timeDiff] + value * timeDiff
+        uint256 n = _sample.d.mul(lastAccumulator.n).add(
+            _sample.n.mul(lastAccumulator.d).mul(_time.sub(lastSampleTime))
+        );
+        uint256 d = _sample.d.mul(lastAccumulator.d);
+        (n, d) = Math.reducedRatio(n, d, MAX_UINT128);
+
+        tokenData.accumulators[_time] = Fraction({ n: n, d: d });
+        tokenData.timestamps[_time] = true;
+        tokenData.lastSampleTime = _time;
+
+        emit SampleAdded(_token, _sample.n, _sample.d, _time);
+    }
+
+    /**
+     * @dev checks whether the accumulator is initialized for a specific token
+     *
+     * @param _token the token the data is accumulated for
+     *
+     * @return whether the accumulator is initialized for a specific token
+     */
+    function isInitialized(IERC20Token _token) private view returns (bool) {
+        TokenData storage tokenData = data[_token];
+        uint256 firstSampleTime = tokenData.firstSampleTime;
+
+        Fraction memory firstAccumulator = tokenData.accumulators[firstSampleTime];
+        return firstSampleTime > 0 && firstAccumulator.n == INITIAL_SAMPLE_N && firstAccumulator.d == INITIAL_SAMPLE_D;
```

Comment:
Is it possible for `firstAccumulator.n` and `firstAccumulator.d` to be anything other than 0 when `firstSampleTime` is 0?
Otherwise, it might be useful to avoid the related storage read in this case.

Reply:
Nope, they shouldn't. I'll remove it